applications/nrf_desktop/src/modules/ble_qos.c

```diff
@@ -0,0 +1,689 @@
+/*
+ * Copyright (c) 2018 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <stdio.h>
+#include <zephyr.h>
+#include <zephyr/types.h>
+#include <device.h>
+#include <drivers/uart.h>
+
+#include <bluetooth/bluetooth.h>
+#include <bluetooth/conn.h>
+
+#ifdef CONFIG_BT_LL_NRFXLIB
+#include "ble_controller_hci_vs.h"
+#else
+#error NRFXLIB is required for channel map management
+#endif
+
+#include "chmap_filter.h"
+
+#define MODULE ble_qos
+#include "module_state_event.h"
+#include "ble_event.h"
+#include "config_event.h"
+#include "hid_event.h"
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(MODULE, CONFIG_DESKTOP_BLE_QOS_LOG_LEVEL);
+
+K_SEM_DEFINE(sem_initialized, 0, 1);
+
+struct params_ble {
+	u16_t sample_count_min;
+	u8_t min_channel_count;
+	s16_t weight_crc_ok;
+	s16_t weight_crc_error;
+	u16_t ble_block_threshold;
+	u8_t eval_max_count;
+	u16_t eval_duration;
+	u16_t eval_keepout_duration;
+	u16_t eval_success_threshold;
+} __packed;
+
+struct params_wifi {
+	s16_t wifi_rating_inc;
+	s16_t wifi_present_threshold;
+	s16_t wifi_active_threshold;
+} __packed;
+
+struct params_chmap {
+	u8_t chmap[5];
+} __packed;
+
+struct params_blacklist {
+	u16_t wifi_chn_bitmask;
+} __packed;
+
+static struct chmap_instance *mp_chmap_inst;
+static u8_t m_current_chmap[5] = {0xFF, 0xFF, 0xFF, 0xFF, 0x1F};
+static atomic_t m_processing;
+static atomic_t m_new_blacklist;
+static struct chmap_filter_params *m_new_params;
+static struct k_mutex m_data_access_mutex;
+
+#if CONFIG_DESKTOP_BLE_QOS_STATS_PRINTOUT_ENABLE
+# ifndef ROUNDED_DIV
+# define ROUNDED_DIV(a, b)  (((a) + ((b) / 2)) / (b))
+# endif /* ROUNDED_DIV */
+# if DT_NORDIC_NRF_USBD_USBD_0_NUM_IN_ENDPOINTS < 4
+# error Too few USB IN Endpoints enabled.\
+Modify appropriate dts.overlay to increase num-in-endpoints to 4 or more
+# endif
+struct device *cdc_dev;
+static u32_t cdc_dtr;
+
+static void ble_chn_stats_print(bool update_channel_map)
+{
+	char str[64];
+	int str_len, sent_len;
+	s16_t chn_rating;
+	u8_t chn_freq, chn_state;
+	bool buf_loss = false;
+
+# if CONFIG_UART_LINE_CTRL
+	/* Repeated to monitor CDC state */
+	uart_line_ctrl_get(cdc_dev, LINE_CTRL_DTR, &cdc_dtr);
+# endif
+
+	if (!cdc_dtr) {
+		return;
+	}
+
+	if (update_channel_map) {
+		str_len = snprintf(
```

Comment:
Be aware. The length returned by `snprintf()` might not be what you think it is. From the [manual](https://linux.die.net/man/3/snprintf):

> If the output was truncated due to this limit then the return value is the number of characters (excluding the terminating null byte) which would have been written to the final string if enough space had been available. Thus, a return value of size **or more** means that the output was truncated.

The following code illustrates the problem:

```c
#include <stdio.h>
int main(int argc, char * argv[])
{
    char buf[4];
    size_t size = snprintf(buf, sizeof(buf),
                           "%s", "This string is too long\n");
    printf("Size %zu\n", size);
    return 0;
}
```
The output of this code with GCC 9.1.0 is `24`.


NOTE: The `snprintf()` implementation in the standard library linked by Zephyr might behave differently. In any case, you need to sanitize the return carefully.

Reply:
Thanks! Will use strlen after snprintf to determine length instead.

Reply:
The logic behind this buffer handling is to fill the 64-byte string buffer as much as possible before transmitting it over USB (USB interrupt endpoint size is 64 bytes). When the buffer reaches 2/3 full in the loop, it's transmitted. Using strlen() would be useful for determining when that 2/3 watermark has been reached